org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/JavaCodeMiningReconciler.java

```diff
@@ -92,7 +103,6 @@ public void uninstall() {
 	}
 
 	public static boolean isReconciled(ISourceViewerExtension5 viewer) {
-		return reconciledViewers.contains(viewer);
+		return !toBeReconciledViewers.containsKey(viewer);
```

Comment:
This would also be a computeifabsent, then checking if the future was completed.

That way there is always one future holding the last state and we can check it.

Reply:
> If we keep the futures forever in the map, we need to find a time when to remove the entry from the map. 

The time is actually when uninstalling the viewer.

> Please again convince me if you think that the current implementation is not correct.

Se previous comments. The current logic is all prone to timing issues and as code mining (or even reconcilation!) can happen on different threads any time and might finish at random times.

Therefore there should only be two states:

1. The future is already completed because reconciliation has happened before
2. The future is currently pending because reconciliation is running.

and all this must be synchronized by the concurrent map.